File: app/src/main/java/com/ducktapedapps/updoot/api/redditAPI.java
package com.ducktapedapps.updoot.api;

import com.ducktapedapps.updoot.model.account;
import com.ducktapedapps.updoot.model.thing;

import io.reactivex.Completable;
import io.reactivex.Single;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface redditAPI {

    @GET("/api/v1/me")
    Single<account> getUserIdentity();

    @GET("{subreddit}/{sort}")
    Single<thing> getSubreddit(
            @Path("subreddit") String subreddit,
            @Path("sort") String sort,
            @Query("t") String time,
            @Query("after") String after);

    @FormUrlEncoded
    @POST("/api/vote")
    Completable castVote(
            @Field("id") String thing_id,
            @Field("dir") int vote_direction
    );

    @FormUrlEncoded
    @POST("/api/save")
    Completable save(
            @Field("id") String id
    );

    @FormUrlEncoded
    @POST("/api/unsave")
    Completable unsave(
            @Field("id") String id
    );

    @GET("{subreddit}/comments/{id}")
    Single<thing> getComments(
            @Path("subreddit") String subreddit,
            @Path("id") String submissions_id
    );

}

File: app/src/main/java/com/ducktapedapps/updoot/api/RedditAPI.kt
package com.ducktapedapps.updoot.api

import com.ducktapedapps.updoot.model.Account
import com.ducktapedapps.updoot.model.CommentListing
import com.ducktapedapps.updoot.model.SearchListing
import com.ducktapedapps.updoot.model.SubmissionListing
import retrofit2.http.*

interface RedditAPI {

    @GET("/api/v1/me")
    suspend fun userIdentity(): Account

    @GET("{subreddit}/{sort}")
    suspend fun getSubreddit(
            @Path("subreddit") subreddit: String?,
            @Path("sort") sort: String,
            @Query("t") time: String?,
            @Query("after") after: String?): SubmissionListing

    @FormUrlEncoded
    @POST("/api/save")
    suspend fun save(
            @Field("id") id: String
    ): String

    @FormUrlEncoded
    @POST("/api/unsave")
    suspend fun unsave(
            @Field("id") id: String
    ): String

    @GET("{subreddit}/comments/{id}")
    suspend fun getComments(
            @Path("subreddit") subreddit: String,
            @Path("id") submissions_id: String
    ): CommentListing

    @GET("/api/morechildren")
    suspend fun getMoreChildren(
            @Query("children") children: String,
            @Query("link_id") link_id: String,
            @Query("api_type") type: String = "json"
    ): String

    @FormUrlEncoded
    @POST("/api/vote")
    suspend fun castVote(
            @Field("id") thing_id: String,
            @Field("dir") vote_direction: Int
    ): String

    @GET("/subreddits/search")
    suspend fun search(
            @Query("q") query: String
    ): SearchListing?
}

